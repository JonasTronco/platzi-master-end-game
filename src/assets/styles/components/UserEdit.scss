@import '../main.scss';

.user-edit {

  &__title {
    @include flexStaCen;
    margin-bottom: $space_x6;

    &__icon {
      width: $space_x8;
      height: $space_x8;
      fill: $secondary_color;
    }

    &__text {
      margin-left: $space_x2;
    }
  }

  &__form {

    &__input {
      margin-bottom: $space_x3;
    }

    &__button {
      margin-top: $space_x3;
    }
  }
}