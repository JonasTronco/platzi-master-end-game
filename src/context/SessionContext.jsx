import { useState, useEffect, createContext, useMemo, useContext } from 'react';
import Router from 'next/router';

const SessionContext = createContext();
export function SessionProvider(props) {
  const [session, setSession] = useState(null);

  useEffect(() => {
    try {
      setSession(JSON.parse(localStorage.getItem('session')));
    } catch (error) {
      Router.push('/');
    }
  }, []);

  const value = useMemo(
    () => ({
      session,
    }),
    [session],
  );

  return <SessionContext.Provider value={value} {...props} />;
}

export function useSession() {
  const context = useContext(SessionContext);
  if (!context) {
    throw new Error('No Session Context Provider');
  }
  return context;
}
