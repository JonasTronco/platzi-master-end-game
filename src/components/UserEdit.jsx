import { useState } from 'react';
import { Formik, Form, Field, ErrorMessage } from 'formik';
import UserIcon from '../assets/icons/user.svg';
import '../assets/styles/components/UserEdit.scss';

function UserEdit() {
  // eslint-disable-next-line prefer-const
  let [status, setStatus] = useState({
    loading: false,
    error: false,
    success: false,
  });

  if (status.loading) {
    return <h3>Loading...</h3>;
  }

  if (status.error) {
    return <h3>Sorry we could not update the user info, please try again</h3>;
  }

  if (status.success) {
    return <h3>User info updated successfully</h3>;
  }

  return (
    <div className="user-edit">
      <div className="user-edit__title">
        <UserIcon className="user-edit__title__icon" />
        <h2 className="user-edit__title__text">Edit User</h2>
      </div>
      <Formik
        initialValues={{ profileImage: '', email: '', phoneNumber: '' }}
        validate={(values) => {
          const errors = {};
          if (!values.profileImage) {
            errors.profileImage = 'Required';
          }
          if (!values.email) {
            errors.email = 'Required';
          }
          if (!values.phoneNumber) {
            errors.phoneNumber = 'Required';
          }
          return errors;
        }}
        onSubmit={(values, { setSubmitting }) => {
          const result = new Promise((res, rej) => {
            setTimeout(() => {
              res(values);
            }, 400);
          });
          result.then((data) => {
            alert(JSON.stringify(data, null, 2));
            setSubmitting(false);
            setStatus(status = {
              loading: false,
              error: false,
              success: true,
            });
          })
            .catch(() => {
              setStatus(status = {
                loading: false,
                error: true,
                success: false,
              });
            });
          setStatus(status = {
            loading: true,
            error: false,
            success: false,
          });
        }}
      >
        {({ isSubmitting }) => (
          <Form className="user-edit__form">
            <div className="user-edit__form__input input lg">
              <strong className="user-edit__form__input__text input__label">Profile Image</strong>
              <Field
                className="user-edit__form__input__field input__field lg"
                type="profileImage"
                name="profileImage"
              />
              <ErrorMessage
                className="user-edit__form__input__error input__error"
                component="div"
                name="profileImage"
              />
            </div>
            <div className="user-edit__form__input input lg">
              <strong className="user-edit__form__input__text input__label">Email</strong>
              <Field
                className="user-edit__form__input__field input__field lg"
                type="email"
                name="email"
              />
              <ErrorMessage
                className="user-edit__form__input__error input__error"
                component="div"
                name="email"
              />
            </div>
            <div className="user-edit__form__input input lg">
              <strong className="user-edit__form__input__text input__label">Phone Number</strong>
              <Field
                className="user-edit__form__input__field input__field lg"
                type="phoneNumber"
                name="phoneNumber"
              />
              <ErrorMessage
                className="user-edit__form__input__error input__error"
                component="div"
                name="phoneNumber"
              />
            </div>
            <button
              className="user-edit__form__button btn lg"
              type="submit"
              disabled={isSubmitting}
            >
              Update user
            </button>
          </Form>
        )}
      </Formik>
    </div>
  );
}

export default UserEdit;
