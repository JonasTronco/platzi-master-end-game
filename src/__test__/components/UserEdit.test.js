import UserEdit from '../../components/UserEdit';
import { create } from 'react-test-renderer';
import { SessionProvider } from '../../context/SessionContext';
describe('UserEdit Component', () => {
  test('Should render UserEdit component', () => {
    const component = create(
      <SessionProvider>
        <UserEdit />
      </SessionProvider>
    );
    expect(component.toJSON()).toMatchSnapshot();
  });
});
