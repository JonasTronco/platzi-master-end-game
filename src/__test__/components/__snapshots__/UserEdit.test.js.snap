// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserEdit Component Should Snapshot UserEdit component 1`] = `
<div
  className="user-edit"
  data-testid="user-edit"
>
  <div
    className="user-edit__title"
  >
    <div
      className="user-edit__title__icon"
    />
    <h2
      className="user-edit__title__text"
    >
      Edit User
    </h2>
  </div>
  <form
    action="#"
    className="user-edit__form"
    onReset={[Function]}
    onSubmit={[Function]}
  >
    <div
      className="user-edit__form__input input lg"
    >
      <strong
        className="user-edit__form__input__text input__label"
      >
        Profile Image
      </strong>
      <input
        className="user-edit__form__input__field input__field lg"
        name="profileImage"
        onBlur={[Function]}
        onChange={[Function]}
        type="profileImage"
        value=""
      />
    </div>
    <div
      className="user-edit__form__input input lg"
    >
      <strong
        className="user-edit__form__input__text input__label"
      >
        Email
      </strong>
      <input
        className="user-edit__form__input__field input__field lg"
        name="email"
        onBlur={[Function]}
        onChange={[Function]}
        type="email"
        value=""
      />
    </div>
    <div
      className="user-edit__form__input input lg"
    >
      <strong
        className="user-edit__form__input__text input__label"
      >
        Phone Number
      </strong>
      <input
        className="user-edit__form__input__field input__field lg"
        name="phoneNumber"
        onBlur={[Function]}
        onChange={[Function]}
        type="phoneNumber"
        value=""
      />
    </div>
    <button
      className="user-edit__form__button btn lg"
      disabled={false}
      type="submit"
    >
      Update user
    </button>
  </form>
</div>
`;
